### ISSUE TITLE
[question] Do you renormalize each step's decoding score during inference?

### ISSUE THREAD

**Issue by entslscheia** [ElementAI/picard#48]
Hi, I have a quick question regarding your constrained decoding process. 

Based on my understanding, the core idea of Picard is to directly impose some constraints for each decoding step (i.e., dynamically deriving a set of admissible tokens) based on the decoding history using T5. My question is, do you normalize the score for these admissible tokens? Or do you use the original probability over the original vocabulary and simply ignore the invalid tokens? In terms of implementation, the former case is basically passing a vocab mask with all invalid tokens being masked out right before doing `softmax`, while the later passing the vocab mask right after `softmax`. Which one exactly do you use? (sorry I did not get enough time to check your code carefully :confused:)

My guess is that you don't actually normalize the score, because intuitively I feel there might be some problems if normalizing them. For example, if for a certain decoding step, there is only one admissible token, then no matter it matches the question or not, it will get a probability of 1. This is just an extreme case to show the problem, but actually similar issues can happen in many general cases.

Could you give me a hint on this? Many thanks!

**Comment by tscholak** [comment]
Hi @entslscheia, the quick answer is: no normalization.

> dynamically deriving a set of admissible tokens

This is not what actually happens. Picard doesn't compute that set. Instead, proposals are presented to Picard, and those are either accepted or rejected. You could build the set of admissible tokens by proposing all tokens in the vocabulary, but that would wasteful. Instead, only the top-k tokens that continue a given hypothesis are checked. k is essentially a branching factor for the beam search. I found that k = 2 works well on Spider and CoSQL with T5.

**Comment by entslscheia** [comment]
> Hi @entslscheia, the quick answer is: no normalization.
> 
> > dynamically deriving a set of admissible tokens
> 
> This is not what actually happens. Picard doesn't compute that set. Instead, proposals are presented to Picard, and those are either accepted or rejected. You could build the set of admissible tokens by proposing all tokens in the vocabulary, but that would be wasteful. Instead, only the top-k tokens that continue a given hypothesis are checked. k is essentially a branching factor for the beam search. I found that k = 2 works well on Spider and CoSQL with T5.

Really appreciate your response! 

I guess now I have a better understanding of Picard. It's really a smart use of pre-trained encoders-decoders. However, I have a following up question as to _top-k tokens that continue a given hypothesis are checked_. Is it possible that all k tokens in the beam are not admissible and thus being rejected, especially when k is small (e.g., k=2)? While this case will not happen if the admissible tokens are derived in advance. Do you agree with that? I am not entirely sure.

**Comment by tscholak** [comment]
> Is it possible that all k tokens in the beam are not admissible and thus being rejected

Yes, that's possible. When neither of the `k = 2` tokens are admissible, then the hypothesis is not continued. Empirically I found that, on Spider and with T5, this does not occur very often. On other datasets, with weaker models, or OOD models that are confidently wrong, this may indeed pose a problem.

There is no guarantee that the model will produce eventually something that parses as a whole expression. Sometimes generation enters a vicious loop where the same nested construct is repeated over and over again until production is cancelled due to timeout or length. Like: `SELECT foo.id FROM foo WHERE (SELECT foo.id FROM foo WHERE (SELECT foo.id FROM foo WHERE (` In such cases, the mistake happens early on, but the model is stuck forever because it has committed to that mistake. These branches typically die quickly though if there are alternatives.

> While this case will not happen if the admissible tokens are derived in advance.

In the limit `k -> infty`, I think most of what I described above will still occur.
